k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapFolder.java

```diff
@@ -1248,17 +1248,40 @@ public String getUidFromMessageId(Message message) throws MessagingException {
     }
 
     @Override
-    public void expunge() throws MessagingException {
+    public void expunge(List<Long> knownDeletedUids) throws MessagingException {
+        if (knownDeletedUids == null || knownDeletedUids.isEmpty()) {
+            return;
+        }
         open(OPEN_MODE_RW);
         checkOpen();
 
+        Set<Long> knownDeletedUidSet = new HashSet<>(knownDeletedUids);
+        //Ensure that we expunge only those messages that we have explicitly marked with the \Deleted flag
+        //For more information, see section 4.2.4 of RFC 4549
         try {
-            executeSimpleCommand("EXPUNGE");
+            if (connection.isUidPlusCapable()) {
+                connection.executeCommandWithIdSet(Commands.UID_EXPUNGE, "", knownDeletedUidSet);
+            } else {
+                expungeWithoutUidPlus(knownDeletedUidSet);
+            }
         } catch (IOException ioe) {
             throw ioExceptionHandler(connection, ioe);
         }
     }
 
+    private void expungeWithoutUidPlus(Set<Long> knownDeletedUids) throws IOException, MessagingException {
```

Comment:
I know this is what RFC 4549 suggests when `UID EXPUNGE` is not available, but I don't particularly like it. It has the potential to undelete messages when the connection is lost in the middle of the command sequence. It also requires passing all known deleted UIDs to the method.

By now UIDPLUS and with that `UID EXPUNGE` should be widely available. So I'm fine with K-9 Mail not making the best effort when using legacy servers. Also, most users don't care about messages marked as deleted. So I'd say fall back to a plain `EXPUNGE` when `UID EXPUNGE` is not available. That is what we're doing right now, anyway.